File: EtherCatMqttGateway/SlaveDevice.cs
using System.Buffers.Binary;
using System.Text;
using EtherCAT.NET;
using EtherCAT.NET.Infrastructure;
using Newtonsoft.Json.Linq;

namespace EtherCatMqttGateway;

/// <summary>
/// Wraps an EtherCAT slave and exposes PDO variables as typed helpers.
/// Handles bit-level read/write and JSON conversions.
/// </summary>
public class SlaveDevice
{
    private readonly EcMaster _master;
    private readonly SlaveInfo _slaveInfo;
    private readonly ushort _slaveIndex; // ring order (1-based), used for MQTT CSA topics
    private readonly List<SlaveVariable> _variables = new();

    /// <summary>
    /// Creates a new slave device wrapper.
    /// </summary>
    /// <param name="master">EtherCAT master.</param>
    /// <param name="slaveInfo">Discovered slave info (from scan).</param>
    /// <param name="slaveIndex">Ring order starting at 1 (used as CSA/topic id).</param>
    public SlaveDevice(EcMaster master, SlaveInfo slaveInfo, ushort slaveIndex)
    {
        _master = master;
        _slaveInfo = slaveInfo;
        _slaveIndex = slaveIndex;

        foreach (var pdo in _slaveInfo.DynamicData.Pdos)
            _variables.AddRange(pdo.Variables);
    }

    /// <summary>Human-readable device name.</summary>
    public string GetName() => _slaveInfo.DynamicData.Name;

    /// <summary>
    /// Controller Station Address used for MQTT topics (ring index supplied by the controller).
    /// </summary>
    public ushort GetCsa() => _slaveIndex;

    /// <summary>
    /// CSA reported by the EtherCAT stack (may or may not match ring index).
    /// </summary>
    public ushort GetReportedCsa() => _slaveInfo.Csa;

    /// <summary>Returns the raw slave info.</summary>
    public SlaveInfo GetSlaveInfo() => _slaveInfo;

    /// <summary>
    /// Metadata for this slave (variables & attributes) as JSON.
    /// </summary>
    public JObject GetMetadata()
    {
        var pdos = new JArray();
        foreach (var variable in _variables)
        {
            var pdo = new JObject
            {
                ["name"] = variable.Name,
                ["bitLength"] = variable.BitLength,
                ["dataType"] = variable.DataType.ToString(),
                ["dataDirection"] = variable.DataDirection.ToString(),
                ["index"] = variable.Index.ToString("X4"),
                ["subIndex"] = variable.SubIndex.ToString("X2")
            };
            pdos.Add(pdo);
        }

        return new JObject
        {
            ["name"] = _slaveInfo.DynamicData.Name,
            ["description"] = _slaveInfo.DynamicData.Description,
            ["reportedCsa"] = _slaveInfo.Csa,
            ["ringCsa"] = _slaveIndex,
            ["pdos"] = pdos
        };
    }

    /// <summary>
    /// Whether the variable has a valid mapped data pointer.
    /// </summary>
    private static bool HasValidPtr(SlaveVariable v) => v.DataPtr != IntPtr.Zero;

    /// <summary>All mapped variables with valid pointers.</summary>
    public List<SlaveVariable> GetAllVariables() => _variables.Where(HasValidPtr).ToList();

    /// <summary>Output variables (Tx to slaves) with valid pointers.</summary>
    public List<SlaveVariable> GetOutputVariables() =>
        _variables.Where(v => v.DataDirection == DataDirection.Output && HasValidPtr(v)).ToList();

    /// <summary>Input variables (Rx from slaves) with valid pointers.</summary>
    public List<SlaveVariable> GetInputVariables() =>
        _variables.Where(v => v.DataDirection == DataDirection.Input && HasValidPtr(v)).ToList();

    /// <summary>
    /// Converts a bit-packed buffer to an unsigned integer (up to 64 bits, LSB first).
    /// </summary>
    private static ulong BufferToUInt(ReadOnlySpan<byte> buffer, byte bits)
    {
        if (bits is < 1 or > 64) throw new ArgumentOutOfRangeException(nameof(bits));
        ulong result = 0;

        for (int i = 0; i < bits; i++)
        {
            int byteIndex = i / 8;
            int bitIndex = i % 8;
            if ((buffer[byteIndex] & (1 << bitIndex)) != 0)
                result |= 1UL << i;
        }
        return result;
    }

    /// <summary>
    /// Converts a bit-packed buffer to a signed integer (up to 64 bits, LSB first).
    /// </summary>
    private static long BufferToInt(ReadOnlySpan<byte> buffer, byte bits)
    {
        ulong u = BufferToUInt(buffer, bits);
        long s = unchecked((long)u);
        if (bits < 64 && ((1L << (bits - 1)) & s) != 0)
            s |= -1L << bits; // sign-extend
        return s;
    }

    /// <summary>
    /// Packs an unsigned integer into a bit-packed buffer (LSB first).
    /// </summary>
    private static byte[] UIntToBuffer(ulong value, byte bits)
    {
        if (bits is < 1 or > 64) throw new ArgumentOutOfRangeException(nameof(bits));
        var buf = new byte[(bits + 7) / 8];
        for (int i = 0; i < bits; i++)
        {
            if ((value & (1UL << i)) != 0)
                buf[i / 8] |= (byte)(1 << (i % 8));
        }
        return buf;
    }

    /// <summary>
    /// Packs a signed integer into a bit-packed buffer (LSB first).
    /// </summary>
    private static byte[] IntToBuffer(long value, byte bits) => UIntToBuffer(unchecked((ulong)value), bits);

    /// <summary>
    /// Reads the variable into a tightly-sized bit-packed buffer (LSB-first).
    /// </summary>
    public byte[]? ReadVariable(SlaveVariable variable)
    {
        var buffer = new byte[(variable.BitLength + 7) / 8];

        unsafe
        {
            var basePtr = (byte*)variable.DataPtr.ToPointer();
            if (basePtr == null) return null;

            int byteOffset = variable.BitOffset / 8;
            int bitOffset = variable.BitOffset % 8;
            int bitLength = variable.BitLength;

            // Copy bit-by-bit from mapped process memory to our compact buffer.
            for (int i = 0; i < bitLength; i++)
            {
                int srcByte = (bitOffset + i) / 8;
                int srcBit = (bitOffset + i) % 8;

                if ((basePtr[byteOffset + srcByte] & (1 << srcBit)) != 0)
                    buffer[i / 8] |= (byte)(1 << (i % 8));
            }
        }
        return buffer;
    }

    /// <summary>
    /// Writes the given bit-packed buffer into the mapped variable (LSB-first).
    /// </summary>
    public void WriteVariable(SlaveVariable variable, ReadOnlySpan<byte> value)
    {
        unsafe
        {
            var basePtr = (byte*)variable.DataPtr.ToPointer();
            if (basePtr == null) throw new InvalidOperationException("Data pointer is null");

            int byteOffset = variable.BitOffset / 8;
            int bitOffset = variable.BitOffset % 8;
            int bitLength = variable.BitLength;

            for (int i = 0; i < bitLength; i++)
            {
                int dstByte = (bitOffset + i) / 8;
                int dstBit = (bitOffset + i) % 8;

                if ((value[i / 8] & (1 << (i % 8))) != 0)
                    basePtr[byteOffset + dstByte] |= (byte)(1 << dstBit);
                else
                    basePtr[byteOffset + dstByte] &= (byte)~(1 << dstBit);
            }
        }
    }

    /// <summary>
    /// Reads a variable and converts it to a JSON token (numbers/strings/byte arrays).
    /// </summary>
    public JToken ReadVariableAsJToken(SlaveVariable variable)
    {
        var buffer = ReadVariable(variable);
        if (buffer == null) return JValue.CreateNull();

        var span = buffer.AsSpan();

        switch (variable.DataType)
        {
            case EthercatDataType.Boolean:
                return new JValue((BufferToUInt(span, variable.BitLength) & 1UL) == 1UL);

            case EthercatDataType.BIT2:
            case EthercatDataType.BIT3:
            case EthercatDataType.BIT4:
            case EthercatDataType.BIT5:
            case EthercatDataType.BIT6:
            case EthercatDataType.BIT7:
            case EthercatDataType.BIT8:
            case EthercatDataType.BITARR8:
            case EthercatDataType.BITARR16:
            case EthercatDataType.BITARR32:
                return new JValue(BufferToUInt(span, variable.BitLength));

            case EthercatDataType.Unsigned8:
                return new JValue((byte)BufferToUInt(span, 8));
            case EthercatDataType.Unsigned16:
                return new JValue(BinaryPrimitives.ReadUInt16LittleEndian(span));
            case EthercatDataType.Unsigned24:
                return new JValue((uint)BufferToUInt(span, 24));
            case EthercatDataType.Unsigned32:
                return new JValue(BinaryPrimitives.ReadUInt32LittleEndian(span));
            case EthercatDataType.Unsigned40:
            case EthercatDataType.Unsigned48:
            case EthercatDataType.Unsigned56:
            case EthercatDataType.Unsigned64:
                return new JValue(BufferToUInt(span, variable.BitLength));

            case EthercatDataType.Integer8:
                return new JValue(unchecked((sbyte)BufferToInt(span, 8)));
            case EthercatDataType.Integer16:
                return new JValue(unchecked((short)BinaryPrimitives.ReadUInt16LittleEndian(span)));
            case EthercatDataType.Integer24:
                return new JValue(unchecked((int)BufferToInt(span, 24)));
            case EthercatDataType.Integer32:
                return new JValue(unchecked((int)BinaryPrimitives.ReadUInt32LittleEndian(span)));
            case EthercatDataType.Integer40:
            case EthercatDataType.Integer48:
            case EthercatDataType.Integer56:
            case EthercatDataType.Integer64:
                return new JValue(BufferToInt(span, variable.BitLength));

            case EthercatDataType.@float:
                // 32-bit IEEE-754 little-endian
                return new JValue(BitConverter.ToSingle(span));
            case EthercatDataType.@double:
                // 64-bit IEEE-754 little-endian
                return new JValue(BitConverter.ToDouble(span));

            case EthercatDataType.VisibleString:
                // Interpret as UTF-8/ASCII until first NUL.
                var nul = Array.IndexOf(buffer, (byte)0);
                var str = Encoding.UTF8.GetString(buffer, 0, nul >= 0 ? nul : buffer.Length);
                return new JValue(str);

            case EthercatDataType.OctetString:
                // Hex representation (safe for dashboards)
                return new JValue(Convert.ToHexString(buffer));

            // TODO: UnicodeString/GUID/Time types as needed
            default:
                return JValue.CreateNull();
        }
    }

    /// <summary>
    /// Converts a JSON token into the on-wire bit-packed representation and writes it.
    /// </summary>
    public void WriteVariableAsJToken(SlaveVariable variable, JToken value)
    {
        switch (variable.DataType)
        {
            case EthercatDataType.Boolean:
                WriteVariable(variable, UIntToBuffer(value.Value<bool>() ? 1UL : 0UL, variable.BitLength));
                return;

            case EthercatDataType.BIT2:
            case EthercatDataType.BIT3:
            case EthercatDataType.BIT4:
            case EthercatDataType.BIT5:
            case EthercatDataType.BIT6:
            case EthercatDataType.BIT7:
            case EthercatDataType.BIT8:
            case EthercatDataType.BITARR8:
            case EthercatDataType.BITARR16:
            case EthercatDataType.BITARR32:
                WriteVariable(variable, UIntToBuffer(value.Value<ulong>(), variable.BitLength));
                return;

            case EthercatDataType.Unsigned8:
            case EthercatDataType.Unsigned16:
            case EthercatDataType.Unsigned24:
            case EthercatDataType.Unsigned32:
            case EthercatDataType.Unsigned40:
            case EthercatDataType.Unsigned48:
            case EthercatDataType.Unsigned56:
            case EthercatDataType.Unsigned64:
                WriteVariable(variable, UIntToBuffer(value.Value<ulong>(), variable.BitLength));
                return;

            case EthercatDataType.Integer8:
            case EthercatDataType.Integer16:
            case EthercatDataType.Integer24:
            case EthercatDataType.Integer32:
            case EthercatDataType.Integer40:
            case EthercatDataType.Integer48:
            case EthercatDataType.Integer56:
            case EthercatDataType.Integer64:
                WriteVariable(variable, IntToBuffer(value.Value<long>(), variable.BitLength));
                return;

            case EthercatDataType.@float:
                {
                    var f = value.Value<float>();
                    Span<byte> tmp = stackalloc byte[4];
                    BinaryPrimitives.WriteSingleLittleEndian(tmp, f);
                    WriteVariable(variable, tmp);
                    return;
                }
            case EthercatDataType.@double:
                {
                    var d = value.Value<double>();
                    Span<byte> tmp = stackalloc byte[8];
                    BinaryPrimitives.WriteDoubleLittleEndian(tmp, d);
                    WriteVariable(variable, tmp);
                    return;
                }

            case EthercatDataType.VisibleString:
                {
                    var s = value.Value<string>() ?? string.Empty;
                    var bytes = Encoding.UTF8.GetBytes(s);
                    WriteVariable(variable, bytes);
                    return;
                }

            case EthercatDataType.OctetString:
                {
                    // Accept hex string (no spaces); otherwise treat as base64 if it looks like it.
                    var str = value.Value<string>() ?? "";
                    byte[] bytes;
                    if (str.Length % 2 == 0 && str.All(Uri.IsHexDigit))
                        bytes = Convert.FromHexString(str);
                    else
                        bytes = Convert.TryFromBase64String(str, new Span<byte>(new byte[str.Length]), out _)
                            ? Convert.FromBase64String(str)
                            : Encoding.UTF8.GetBytes(str);
                    WriteVariable(variable, bytes);
                    return;
                }

            default:
                // Silently ignore unsupported/opaque types for now.
                return;
        }
    }

    /// <summary>
    /// Legacy strongly-typed getters: kept for compatibility; prefer JSON APIs.
    /// </summary>
    public object? GetVariableValue(SlaveVariable variable)
    {
        if (variable.DataType == 0) return null;
        var buf = ReadVariable(variable);
        if (buf == null) return null;

        var span = buf.AsSpan();

        switch (variable.DataType)
        {
            case EthercatDataType.Boolean:
                return (BufferToUInt(span, variable.BitLength) & 1UL) == 1UL;

            case EthercatDataType.BIT2:
            case EthercatDataType.BIT3:
            case EthercatDataType.BIT4:
            case EthercatDataType.BIT5:
            case EthercatDataType.BIT6:
            case EthercatDataType.BIT7:
            case EthercatDataType.BIT8:
                // Return packed bits as ulong
                return BufferToUInt(span, variable.BitLength);

            case EthercatDataType.BITARR8:
            case EthercatDataType.Unsigned8:
                return (byte)BufferToUInt(span, 8);

            case EthercatDataType.BITARR16:
            case EthercatDataType.Unsigned16:
                return BinaryPrimitives.ReadUInt16LittleEndian(span);

            case EthercatDataType.BITARR32:
            case EthercatDataType.Unsigned32:
                return BinaryPrimitives.ReadUInt32LittleEndian(span);

            case EthercatDataType.Unsigned64:
                return BinaryPrimitives.ReadUInt64LittleEndian(span);

            case EthercatDataType.@float:
                return BitConverter.ToSingle(span);

            case EthercatDataType.@double:
                return BitConverter.ToDouble(span);

            case EthercatDataType.Integer8:
                return unchecked((sbyte)BufferToInt(span, 8));

            case EthercatDataType.Integer16:
                return unchecked((short)BinaryPrimitives.ReadUInt16LittleEndian(span));

            case EthercatDataType.Integer32:
                return unchecked((int)BinaryPrimitives.ReadUInt32LittleEndian(span));

            case EthercatDataType.Integer64:
                return unchecked((long)BinaryPrimitives.ReadUInt64LittleEndian(span));

            default:
                throw new ArgumentOutOfRangeException(nameof(variable.DataType), variable.DataType, "Unsupported data type");
        }
    }

    /// <summary>
    /// Legacy strongly-typed setter: prefer WriteVariableAsJToken.
    /// </summary>
    public void SetVariableValue(SlaveVariable variable, object value)
    {
        if (variable.DataType == 0) return;

        switch (variable.DataType)
        {
            case EthercatDataType.Boolean:
                WriteVariable(variable, UIntToBuffer(((bool)value) ? 1UL : 0UL, 1));
                break;

            case EthercatDataType.BIT2:
            case EthercatDataType.BIT3:
            case EthercatDataType.BIT4:
            case EthercatDataType.BIT5:
            case EthercatDataType.BIT6:
            case EthercatDataType.BIT7:
            case EthercatDataType.BIT8:
                WriteVariable(variable, UIntToBuffer(Convert.ToUInt64(value), variable.BitLength));
                break;

            case EthercatDataType.BITARR8:
            case EthercatDataType.Unsigned8:
                WriteVariable(variable, new[] { Convert.ToByte(value) });
                break;

            case EthercatDataType.BITARR16:
            case EthercatDataType.Unsigned16:
                {
                    Span<byte> tmp = stackalloc byte[2];
                    BinaryPrimitives.WriteUInt16LittleEndian(tmp, Convert.ToUInt16(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.BITARR32:
            case EthercatDataType.Unsigned32:
                {
                    Span<byte> tmp = stackalloc byte[4];
                    BinaryPrimitives.WriteUInt32LittleEndian(tmp, Convert.ToUInt32(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.Unsigned64:
                {
                    Span<byte> tmp = stackalloc byte[8];
                    BinaryPrimitives.WriteUInt64LittleEndian(tmp, Convert.ToUInt64(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.@float:
                {
                    Span<byte> tmp = stackalloc byte[4];
                    BinaryPrimitives.WriteSingleLittleEndian(tmp, Convert.ToSingle(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.@double:
                {
                    Span<byte> tmp = stackalloc byte[8];
                    BinaryPrimitives.WriteDoubleLittleEndian(tmp, Convert.ToDouble(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.Integer8:
                WriteVariable(variable, new[] { unchecked((byte)Convert.ToSByte(value)) });
                break;

            case EthercatDataType.Integer16:
                {
                    Span<byte> tmp = stackalloc byte[2];
                    BinaryPrimitives.WriteInt16LittleEndian(tmp, Convert.ToInt16(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.Integer32:
                {
                    Span<byte> tmp = stackalloc byte[4];
                    BinaryPrimitives.WriteInt32LittleEndian(tmp, Convert.ToInt32(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            case EthercatDataType.Integer64:
                {
                    Span<byte> tmp = stackalloc byte[8];
                    BinaryPrimitives.WriteInt64LittleEndian(tmp, Convert.ToInt64(value));
                    WriteVariable(variable, tmp);
                    break;
                }

            default:
                throw new ArgumentOutOfRangeException(nameof(variable.DataType), variable.DataType, "Unsupported data type");
        }
    }
}

File: src/src/slave_device.cpp
#include "ecmqtt/slave_device.hpp"

#include <algorithm>
#include <array>
#include <cctype>
#include <cstdio>
#include <cstring>
#include <stdexcept>

namespace ecmqtt {

namespace {

// Extracts bitLength bits starting at bitOffset (0-7) within base into a
// tightly packed, LSB-first buffer -- mirrors SlaveDevice.cs's ReadVariable.
std::vector<uint8_t> ReadBits(const uint8_t* base, uint8_t bitOffset, uint16_t bitLength) {
    std::vector<uint8_t> buf((bitLength + 7) / 8, 0);
    for (uint16_t i = 0; i < bitLength; ++i) {
        int srcByte = (bitOffset + i) / 8;
        int srcBit = (bitOffset + i) % 8;
        if (base[srcByte] & (1 << srcBit))
            buf[i / 8] |= static_cast<uint8_t>(1 << (i % 8));
    }
    return buf;
}

void WriteBits(uint8_t* base, uint8_t bitOffset, uint16_t bitLength, const uint8_t* value) {
    for (uint16_t i = 0; i < bitLength; ++i) {
        int dstByte = (bitOffset + i) / 8;
        int dstBit = (bitOffset + i) % 8;
        if (value[i / 8] & (1 << (i % 8)))
            base[dstByte] |= static_cast<uint8_t>(1 << dstBit);
        else
            base[dstByte] &= static_cast<uint8_t>(~(1 << dstBit));
    }
}

uint64_t BufferToUInt(const uint8_t* buf, uint16_t bits) {
    uint64_t result = 0;
    for (uint16_t i = 0; i < bits; ++i) {
        int byteIndex = i / 8;
        int bitIndex = i % 8;
        if (buf[byteIndex] & (1 << bitIndex))
            result |= (1ULL << i);
    }
    return result;
}

int64_t BufferToInt(const uint8_t* buf, uint16_t bits) {
    uint64_t u = BufferToUInt(buf, bits);
    int64_t s = static_cast<int64_t>(u);
    if (bits < 64 && ((1LL << (bits - 1)) & s) != 0)
        s |= (-1LL << bits); // sign-extend
    return s;
}

std::vector<uint8_t> UIntToBuffer(uint64_t value, uint16_t bits) {
    std::vector<uint8_t> buf((bits + 7) / 8, 0);
    for (uint16_t i = 0; i < bits; ++i) {
        if (value & (1ULL << i))
            buf[i / 8] |= static_cast<uint8_t>(1 << (i % 8));
    }
    return buf;
}

std::vector<uint8_t> IntToBuffer(int64_t value, uint16_t bits) {
    return UIntToBuffer(static_cast<uint64_t>(value), bits);
}

bool IsHexDigits(const std::string& s) {
    if (s.empty()) return false;
    return std::all_of(s.begin(), s.end(), [](unsigned char c) { return std::isxdigit(c); });
}

std::string ToHexString(const std::vector<uint8_t>& buf) {
    static const char* hex = "0123456789ABCDEF";
    std::string out;
    out.reserve(buf.size() * 2);
    for (uint8_t b : buf) {
        out.push_back(hex[b >> 4]);
        out.push_back(hex[b & 0xF]);
    }
    return out;
}

std::vector<uint8_t> FromHexString(const std::string& s) {
    std::vector<uint8_t> out(s.size() / 2);
    for (size_t i = 0; i < out.size(); ++i)
        out[i] = static_cast<uint8_t>(std::stoul(s.substr(i * 2, 2), nullptr, 16));
    return out;
}

} // namespace

SlaveDevice::SlaveDevice(DiscoveredSlave& slave, std::string name, std::string description)
    : slave_(&slave), name_(std::move(name)), description_(std::move(description)) {}

nlohmann::json SlaveDevice::GetMetadata() const {
    nlohmann::json pdos = nlohmann::json::array();
    for (const auto& v : slave_->variables) {
        char indexHex[8];
        char subHex[8];
        std::snprintf(indexHex, sizeof(indexHex), "%04X", v.index);
        std::snprintf(subHex, sizeof(subHex), "%02X", v.subIndex);

        pdos.push_back({
            {"name", v.name},
            {"description", v.description},
            {"bitLength", v.bitLength},
            {"dataType", ToString(v.dataType)},
            {"dataDirection", ToString(v.direction)},
            {"index", indexHex},
            {"subIndex", subHex},
        });
    }

    return {
        {"name", name_},
        {"description", description_},
        {"state", ToString(slave_->alState)},
        {"reportedCsa", slave_->reportedCsa},
        {"ringCsa", slave_->ringCsa},
        {"pdos", pdos},
    };
}

std::vector<SlaveVariable*> SlaveDevice::GetAllVariables() {
    std::vector<SlaveVariable*> result;
    for (auto& v : slave_->variables)
        if (v.dataPtr != nullptr)
            result.push_back(&v);
    return result;
}

std::vector<SlaveVariable*> SlaveDevice::GetInputVariables() {
    std::vector<SlaveVariable*> result;
    for (auto& v : slave_->variables)
        if (v.dataPtr != nullptr && v.direction == DataDirection::Input)
            result.push_back(&v);
    return result;
}

std::vector<SlaveVariable*> SlaveDevice::GetOutputVariables() {
    std::vector<SlaveVariable*> result;
    for (auto& v : slave_->variables)
        if (v.dataPtr != nullptr && v.direction == DataDirection::Output)
            result.push_back(&v);
    return result;
}

nlohmann::json SlaveDevice::ReadVariableAsJson(const SlaveVariable& variable) const {
    if (variable.dataPtr == nullptr) return nullptr;

    auto buf = ReadBits(variable.dataPtr, variable.bitOffset, variable.bitLength);
    const uint8_t* p = buf.data();

    switch (variable.dataType) {
        case EthercatDataType::Boolean:
            return (BufferToUInt(p, variable.bitLength) & 1ULL) == 1ULL;

        case EthercatDataType::Bit2: case EthercatDataType::Bit3: case EthercatDataType::Bit4:
        case EthercatDataType::Bit5: case EthercatDataType::Bit6: case EthercatDataType::Bit7:
        case EthercatDataType::Bit8:
        case EthercatDataType::BitArr8: case EthercatDataType::BitArr16: case EthercatDataType::BitArr32:
            return BufferToUInt(p, variable.bitLength);

        case EthercatDataType::Unsigned8:
            return static_cast<uint8_t>(BufferToUInt(p, 8));
        case EthercatDataType::Unsigned16: {
            uint16_t v; std::memcpy(&v, p, 2); return v;
        }
        case EthercatDataType::Unsigned24:
            return static_cast<uint32_t>(BufferToUInt(p, 24));
        case EthercatDataType::Unsigned32: {
            uint32_t v; std::memcpy(&v, p, 4); return v;
        }
        case EthercatDataType::Unsigned40: case EthercatDataType::Unsigned48:
        case EthercatDataType::Unsigned56: case EthercatDataType::Unsigned64:
            return BufferToUInt(p, variable.bitLength);

        case EthercatDataType::Integer8:
            return static_cast<int8_t>(BufferToInt(p, 8));
        case EthercatDataType::Integer16: {
            int16_t v; std::memcpy(&v, p, 2); return v;
        }
        case EthercatDataType::Integer24:
            return static_cast<int32_t>(BufferToInt(p, 24));
        case EthercatDataType::Integer32: {
            int32_t v; std::memcpy(&v, p, 4); return v;
        }
        case EthercatDataType::Integer40: case EthercatDataType::Integer48:
        case EthercatDataType::Integer56: case EthercatDataType::Integer64:
            return BufferToInt(p, variable.bitLength);

        case EthercatDataType::Float: {
            float v; std::memcpy(&v, p, 4); return v;
        }
        case EthercatDataType::Double: {
            double v; std::memcpy(&v, p, 8); return v;
        }

        case EthercatDataType::VisibleString: {
            auto nul = std::find(buf.begin(), buf.end(), 0);
            return std::string(buf.begin(), nul);
        }

        case EthercatDataType::OctetString:
            return ToHexString(buf);

        default:
            return nullptr;
    }
}

void SlaveDevice::WriteVariableAsJson(const SlaveVariable& variable, const nlohmann::json& value) {
    if (variable.dataPtr == nullptr) return;

    switch (variable.dataType) {
        case EthercatDataType::Boolean:
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength,
                      UIntToBuffer(value.get<bool>() ? 1ULL : 0ULL, variable.bitLength).data());
            return;

        case EthercatDataType::Bit2: case EthercatDataType::Bit3: case EthercatDataType::Bit4:
        case EthercatDataType::Bit5: case EthercatDataType::Bit6: case EthercatDataType::Bit7:
        case EthercatDataType::Bit8:
        case EthercatDataType::BitArr8: case EthercatDataType::BitArr16: case EthercatDataType::BitArr32:
        case EthercatDataType::Unsigned8: case EthercatDataType::Unsigned16: case EthercatDataType::Unsigned24:
        case EthercatDataType::Unsigned32: case EthercatDataType::Unsigned40: case EthercatDataType::Unsigned48:
        case EthercatDataType::Unsigned56: case EthercatDataType::Unsigned64:
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength,
                      UIntToBuffer(value.get<uint64_t>(), variable.bitLength).data());
            return;

        case EthercatDataType::Integer8: case EthercatDataType::Integer16: case EthercatDataType::Integer24:
        case EthercatDataType::Integer32: case EthercatDataType::Integer40: case EthercatDataType::Integer48:
        case EthercatDataType::Integer56: case EthercatDataType::Integer64:
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength,
                      IntToBuffer(value.get<int64_t>(), variable.bitLength).data());
            return;

        case EthercatDataType::Float: {
            float f = value.get<float>();
            std::array<uint8_t, 4> tmp;
            std::memcpy(tmp.data(), &f, 4);
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength, tmp.data());
            return;
        }
        case EthercatDataType::Double: {
            double d = value.get<double>();
            std::array<uint8_t, 8> tmp;
            std::memcpy(tmp.data(), &d, 8);
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength, tmp.data());
            return;
        }

        case EthercatDataType::VisibleString: {
            std::string s = value.get<std::string>();
            std::vector<uint8_t> bytes(s.begin(), s.end());
            bytes.resize((variable.bitLength + 7) / 8, 0);
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength, bytes.data());
            return;
        }

        case EthercatDataType::OctetString: {
            // Accepts a hex string, else raw UTF-8 bytes (the .NET original also
            // tried base64; omitted here since nothing else in this project needs it).
            std::string s = value.get<std::string>();
            std::vector<uint8_t> bytes;
            if (s.size() % 2 == 0 && IsHexDigits(s))
                bytes = FromHexString(s);
            else
                bytes.assign(s.begin(), s.end());
            bytes.resize((variable.bitLength + 7) / 8, 0);
            WriteBits(variable.dataPtr, variable.bitOffset, variable.bitLength, bytes.data());
            return;
        }

        default:
            // Silently ignore unsupported/opaque types, matching the original.
            return;
    }
}

} // namespace ecmqtt
